src/contexts/WalletContext/services.ts

```diff
@@ -52,6 +52,53 @@ export const checkForRiskyWallet = async (walletAddress: string) => {
   }
 };
 
+// Check if the provided signature is valid for the smart contract at signer.getAddress(),
+// using the checks defined in EIP-1271.
+//
+// For simplicity we use the user's Metamask provider and not Alchemy/Infura, though we could change
+// that in future.
+export const isValidEIP1271Signature = async (
+  signer: ethers.providers.JsonRpcSigner,
+  message: string,
+  signature: string,
+): Promise<boolean> => {
+  const EIP1271_ABI = [
+    'function isValidSignature(bytes32 _message, bytes _signature) public view returns (bytes4)',
+  ];
+
+  // bytes4(keccak256("isValidSignature(bytes32,bytes)")
+  const EIP1271_magic_value = '0x1626ba7e';
+
+  // E.g. 0xd2b098b48bc04cb4a53d8e377e313b70539f402f
+  const walletAddress = await signer.getAddress();
+
+  const contractWallet = new ethers.Contract(walletAddress, EIP1271_ABI, signer);
+  const hashMessage = ethers.utils.hashMessage(message);
+
+  try {
+    const returnValue = await contractWallet.isValidSignature(hashMessage, signature);
+    if (returnValue.toString() === EIP1271_magic_value) {
+      return true;
+    }
+    return false;
+  } catch (error) {
+    // signature is not valid
+    return false;
+  }
+};
+
+// Check if the message has been "signed" on chain by the smart contract at signer.getAddress(),
+// using the interface defined in EIP-1271. (For at least Gnosis, on-chain signature checks will
+// kick in if signature = 0x.)
+export const isAlreadyEIP1271SignedOnChain = async (
```

Comment:
Lets write unit tests for this and keep on improving test coverage of repo 🙏

Reply:
See https://github.com/Voltz-Protocol/voltz-ui/pull/502#discussion_r1071093981